Name: **Rachel Jasmine Canaman** <br>
Section: **DS4A**

<h2 align="center"><b>Laboratory Task 3</b></h2>

<div style="text-align: center;">
    <img src="image2.png" alt="A centered image" style="display: block; margin: 0 auto;">
</div>

**Objective:**  
Perform a forward and backward propagation in Python using the inputs from Laboratory Task 2.

We will: <br>
-Perform a forward pass (already done in Task 2). <br>
-Compute the error and gradients via backward propagation. <br>
-Update weights using gradient descent.

**Step 1: Import Libraries**

In [2]:
import numpy as np

**Step 2: Initialize Data, Weights, and Parameters**

In [3]:
# Input and target
x = np.array([1, 0, 1])   # input vector
y = np.array([1])         # target

# Hidden layer weights
W_hidden = np.array([
    [0.2, -0.3],
    [0.4,  0.1],
    [-0.5, 0.2]
])

# Biases
theta1, theta2, theta3 = -0.4, 0.2, 0.1

# Output weights
w_out = np.array([-0.3, -0.2])

# Learning rate
lr = 0.001

**Step 3: Define Helper Functions** <br>
We use ReLU as activation and its derivative for backpropagation.

In [4]:
# ReLU activation
def relu(z):
    return np.maximum(0, z)

# Derivative of ReLU
def relu_derivative(z):
    return (z > 0).astype(float)

**Step 4: Forward Pass**

In [5]:
# Hidden layer forward pass
z_hidden = x @ W_hidden + np.array([theta1, theta2])
h = relu(z_hidden)

# Output layer forward pass
z_out = h @ w_out + theta3
y_hat = relu(z_out)

# Compute error (MSE with 1/2 factor)
error = 0.5 * (y - y_hat)**2

print("Hidden pre-activations:", z_hidden)
print("Hidden activations:", h)
print("Output pre-activation:", z_out)
print("Predicted output:", y_hat)
print("Error:", error)

Hidden pre-activations: [-0.7  0.1]
Hidden activations: [0.  0.1]
Output pre-activation: 0.08
Predicted output: 0.08
Error: [0.4232]


**Step 5: Backward Pass** <br>
We compute gradients using the chain rule:

1. $\frac{\partial E}{\partial \hat{y}} = \hat{y} - y$
2. Propagate through output layer weights and ReLU derivative.  
3. Propagate through hidden layer weights and ReLU derivative.


In [6]:
# Gradient of error wrt prediction
dE_dyhat = y_hat - y  

# Gradient at output pre-activation
dyhat_dzout = relu_derivative(z_out)  
dE_dzout = dE_dyhat * dyhat_dzout  

# Gradients for output weights and bias
dE_dw_out = h * dE_dzout
dE_dtheta3 = dE_dzout

# Backprop to hidden layer
dE_dh = w_out * dE_dzout
dh_dz = relu_derivative(z_hidden)
dE_dz_hidden = dE_dh * dh_dz

# Gradients for hidden weights and biases
dE_dW_hidden = np.outer(x, dE_dz_hidden)
dE_dtheta_hidden = dE_dz_hidden

print("Gradients for output weights:", dE_dw_out)
print("Gradient for output bias:", dE_dtheta3)
print("Gradients for hidden weights:\n", dE_dW_hidden)
print("Gradients for hidden biases:", dE_dtheta_hidden)

Gradients for output weights: [-0.    -0.092]
Gradient for output bias: [-0.92]
Gradients for hidden weights:
 [[0.    0.184]
 [0.    0.   ]
 [0.    0.184]]
Gradients for hidden biases: [0.    0.184]


**Step 6: Update Weights** <br>

We apply gradient descent: $w \leftarrow w - \eta \cdot \frac{\partial E}{\partial w}$.

In [7]:
# Update output weights and bias
w_out -= lr * dE_dw_out
theta3 -= lr * dE_dtheta3

# Update hidden weights and biases
W_hidden -= lr * dE_dW_hidden
theta1 -= lr * dE_dtheta_hidden[0]
theta2 -= lr * dE_dtheta_hidden[1]

print("Updated hidden weights:\n", W_hidden)
print("Updated hidden biases:", [theta1, theta2])
print("Updated output weights:", w_out)
print("Updated output bias:", theta3)

Updated hidden weights:
 [[ 0.2      -0.300184]
 [ 0.4       0.1     ]
 [-0.5       0.199816]]
Updated hidden biases: [np.float64(-0.4), np.float64(0.19981600000000002)]
Updated output weights: [-0.3      -0.199908]
Updated output bias: [0.10092]


**Conclusion**

- The forward pass produced a prediction of  $\hat{y} \approx 0.08$. 
- The error was relatively high $E \approx 0.423$, consistent with Task 2.  
- Backward propagation computed gradients for all weights and biases.  
- A small learning rate $\eta = 0.001$ was applied to update parameters.  

This exercise demonstrates the mechanics of backpropagation: errors flow backward from the output to the hidden layer, adjusting weights to reduce prediction error. We have successfully implemented both forward and backward propagation. This process underpins all neural network training, enabling learning from data through error minimization.